optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +124,94 @@ private void initBenchmarkDirectoryAndSubdirs() {
 
-    protected long warmUp(ProblemBenchmarkResult problemBenchmarkResult,
-            long startingTimeMillis, long warmUpTimeMillisSpentLimit, long timeLeft) {
-        for (SingleBenchmarkResult singleBenchmarkResult : problemBenchmarkResult.getSingleBenchmarkResultList()) {
-            SolverBenchmarkResult solverBenchmarkResult = singleBenchmarkResult.getSolverBenchmarkResult();
+    private void warmUpStarted(Map<SingleBenchmarkRunner, Future<SingleBenchmarkRunner>> futureMap,
+                               Map<SolverBenchmarkResult, TerminationConfig> originalTerminationConfigMap, long timeLeft) {
+
+        for (SolverBenchmarkResult solverBenchmarkResult : plannerBenchmarkResult.getSolverBenchmarkResultList()) {
             TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
             TerminationConfig tmpTerminationConfig = originalTerminationConfig == null
                     ? new TerminationConfig() : originalTerminationConfig.clone();
             tmpTerminationConfig.shortenTimeMillisSpentLimit(timeLeft);
             solverBenchmarkResult.getSolverConfig().setTerminationConfig(tmpTerminationConfig);
+            originalTerminationConfigMap.put(solverBenchmarkResult, originalTerminationConfig);
+            for (SingleBenchmarkResult singleBenchmarkResult : solverBenchmarkResult.getSingleBenchmarkResultList()) {
+                SingleBenchmarkRunner singleBenchmarkRunner = new SingleBenchmarkRunner(singleBenchmarkResult);
+                Future<SingleBenchmarkRunner> future = warmUpExecutorService.submit(singleBenchmarkRunner);
+                futureMap.put(singleBenchmarkRunner, future);
+            }
+        }
+    }
 
+    private void warmUp(Map<SingleBenchmarkRunner, Future<SingleBenchmarkRunner>> futureMap, long timeEnd) {
+        // Wait for the warm up benchmarks to complete or time-out
+        for (Map.Entry<SingleBenchmarkRunner, Future<SingleBenchmarkRunner>> futureEntry : futureMap.entrySet()) {
+            SingleBenchmarkRunner singleBenchmarkRunner = futureEntry.getKey();
+            Future<SingleBenchmarkRunner> future = futureEntry.getValue();
+            Throwable failureThrowable = null;
             try {
-                Solver solver = solverBenchmarkResult.getSolverConfig().buildSolver();
-                solver.solve(problemBenchmarkResult.readPlanningProblem());
-            } catch (RuntimeException e) {
-                singleBenchmarkRunnerExceptionLogger.warn("The warmUp of singleBenchmark ("
-                        + singleBenchmarkResult.getName() + ") failed.", e);
-                // TODO update firstFailureSingleBenchmarkRunner (when warmups happen in a Runner too in parallel)
+                // Explicitly returning it in the Callable guarantees memory visibility
+                long timeLeft = timeEnd - System.currentTimeMillis();
+                singleBenchmarkRunner = future.get(timeLeft, TimeUnit.MILLISECONDS);
```

Comment:
Suppose for 1 warmup single benchmark we have 10 seconds. If that 10 seconds is set in the terminationConfig (and optaplanner respects that!), why do we need to use that 10 seconds as a timeout too?